### PR TITLE
Unique Spans

### PR DESCRIPTION
**Pull request by crusaderky**
- Part of #7860

Change spans so that, if a user opens and reopens a span context manager with the same tag, it counts as separate spans - e.g. in case of repeated calsl to the same high-level library.

The 'default' span now automatically closes and reopens itself into a new 'default' span when the cluster is idle, save for all tasks not explicitly annotated with `span`. This effectively replicates 1:1 the behaviour of computations.

CC @hendrikmakait @mrocklin @ntabris

### CODE REVIEW

File: distributed/spans.py [hunk 1]
```diff
@@ -36,39 +46,59 @@ def span(*tags: str) -> Iterator[None]:
...
-        yield
+    if not tags:
+        raise ValueError("Must specify at least one span tag")
+
+    prev_tags = dask.config.get("annotations.span.name", ())
+    # You must specify the full history of IDs, not just the parent, because
+    # otherwise you would not be able to uniquely identify grandparents when
+    # they have no tasks of their own.
+    prev_ids = dask.config.get("annotations.span.ids", ())
+    ids = tuple(str(uuid.uuid4()) for _ in tags)
```
Comment (hendrikmakait) [inline distributed/spans.py:81]: Why generate a series of IDs here? If I understand correctly, this span (even with multiple tags) will be nested directly under its parent. Is this for simplicity to not have to deal with cases where a single contextmanager declares multiple tags?

File: distributed/spans.py [hunk 2]
```diff
@@ -205,72 +266,92 @@ def nbytes_total(self) -> int:
...
-        return out
+            # The span may be completely different from the one referenced by the
+            # annotation, due to the TaskGroup collision issue explained above.
+            # Remove the annotation to avoid confusion, and instead rely on
+            # distributed.scheduler.TaskState.group.span_id and
+            # distributed.worker_state_machine.TaskState.span_id.
+            ts.annotations.pop("span", None)
+
+    def _ensure_default_span(self) -> Span:
+        """Return the currently active default span, or create one if the previous one
```
Comment (hendrikmakait) [inline distributed/spans.py:323]: Out-of-scope: I'm not a huge fan of this as it's not very reproducible. Let's keep thinking about a better way to reproduce the progress/code-tracking behavior from computations.

**Review by hendrikmakait — APPROVED** [review]
Thanks, @crusaderky!

File: distributed/spans.py (hunk 1 above)
Comment (crusaderky) [reply distributed/spans.py:81]: Yes, this is to simplify the use case of multiple tags declared at once.

File: distributed/spans.py (hunk 2 above)
Comment (crusaderky) [reply distributed/spans.py:323]: I don't know; I don't hate it now.
If you don't explicitly use spans, you get something that somehow makes sense, at the cost of some mild non-determinism.
If you do use spans, everything is deterministic.

**Pull request merged**